lib/Ast.ml

```diff
@@ -194,19 +157,74 @@ module Indexing_op = struct
         | Some (op, rhs) -> Some {lhs; op; rhs; loc} )
 end
 
-let is_index_op_id ident = Option.is_some (Indexing_op.parse ident)
+module String_id = struct
+  let is_prefix i =
+    match i with
+    | "!=" -> false
+    | _ -> ( match i.[0] with '!' | '?' | '~' -> true | _ -> false )
+
+  let is_monadic_binding s =
+    Int.(String.length s > 3)
+    && (String.is_prefix s ~prefix:"let" || String.is_prefix s ~prefix:"and")
+    && Option.is_none
+         (String.lfindi s ~pos:3 ~f:(fun _ c -> not (Char.is_kwdop c)))
+
+  let is_infix i =
+    if Char.is_infixop i.[0] then true
+    else
+      match i with
+      | "!=" | "land" | "lor" | "lxor" | "mod" | "::" | ":=" | "asr"
+       |"lsl" | "lsr" | "or" | "||" ->
+          true
+      | _ -> is_monadic_binding i
+
+  let is_hash_getter i =
+    let is_infix_char c = Char.equal c '.' || Char.is_infixop c in
+    match (i.[0], i.[String.length i - 1]) with
+    | '#', ('#' | '.') when String.for_all i ~f:is_infix_char -> true
+    | _ -> false
+
+  let is_index_op ident = Option.is_some (Indexing_op.parse ident)
+
+  let is_symbol i = is_prefix i || is_infix i || is_index_op i
+end
+
+module Longident = struct
+  include Longident
 
-let is_index_op exp =
-  match exp.pexp_desc with
-  | Pexp_ident {txt= lident; _} -> (
-    match List.last (Longident.flatten lident) with
-    | Some ident -> is_index_op_id ident
-    | None -> false )
-  | _ -> false
+  let test ~f = function Longident.Lident i -> f i | _ -> false
+
+  let is_prefix = test ~f:String_id.is_prefix
+
+  let is_monadic_binding = test ~f:String_id.is_monadic_binding
+
+  let is_infix = test ~f:String_id.is_infix
 
-let is_symbol_id i = is_prefix_id i || is_infix_id i || is_index_op_id i
+  let is_hash_getter = test ~f:String_id.is_hash_getter
 
-let is_symbol e = is_prefix e || is_infix e || is_index_op e
+  let is_index_op i =
+    match List.last (Longident.flatten i) with
```

Comment:
There is `Longident.last`, which has the advantage of not having the `None` case.

Reply:
done